presto-raptorx/src/main/java/com/facebook/presto/raptorx/metadata/CommitCleaner.java

```diff
@@ -0,0 +1,233 @@
+                .collect(toImmutableList());
+
+        this.clock = requireNonNull(clock, "clock is null");
+    }
+
+    public synchronized void removeOldCommits()
+    {
+        long activeCommitId = transactionManager.oldestActiveCommitId();
+
+        Set<Long> droppedTableIds = masterDao.getDroppedTableIds(activeCommitId);
+
+        Map<Integer, Set<Long>> droppedTableIdsByShard = droppedTableIds.stream()
+                .collect(groupingBy(tableId -> tableShard(tableId, shardDao.size()), toSet()));
+
+        // drop index tables
+        for (long tableId : droppedTableIds) {
+            shardDao.get(tableShard(tableId, shardDao.size())).useHandle(handle ->
+                    handle.execute("DROP TABLE IF EXISTS " + chunkIndexTable(tableId)));
+        }
+
+        // cleanup deleted chunks
+        long now = clock.millis();
+        for (int i = 0; i < shardDao.size(); i++) {
+            ShardCommitCleanerDao shard = shardDao.get(i);
+            Set<Long> tableIds = droppedTableIdsByShard.getOrDefault(i, ImmutableSet.of());
+
+            // cleanup chunks for old commits
+            while (true) {
+                List<TableChunk> chunks = shard.getDeletedChunks(activeCommitId, CHUNK_BATCH_SIZE);
+                if (chunks.isEmpty()) {
+                    break;
+                }
+                cleanupChunks(shard, now, chunks, tableIds);
+            }
+
+            // cleanup chunks for dropped tables
+            if (!tableIds.isEmpty()) {
+                while (true) {
+                    List<TableChunk> chunks = shard.getDeletedChunks(tableIds, CHUNK_BATCH_SIZE);
+                    if (chunks.isEmpty()) {
+                        break;
+                    }
+                    cleanupChunks(shard, now, chunks, tableIds);
+                }
+            }
+        }
+
+        // cleanup dropped table columns
+        if (!droppedTableIds.isEmpty()) {
+            masterDao.cleanupDroppedTableColumns(droppedTableIds);
+        }
+
+        // cleanup schemas, tables, columns, views, commits
+        masterDao.cleanup(activeCommitId);
+
+        // cleanup table sizes
+        for (ShardCommitCleanerDao dao : shardDao) {
+            dao.cleanupTableSizes(activeCommitId);
+        }
+
+        // cleanup successful transactions
+        Set<Long> transactionIds = masterDao.getSuccessfulTransactionIds();
+        for (Iterable<Long> ids : partition(transactionIds, 1000)) {
+            for (ShardCommitCleanerDao shard : shardDao) {
+                shard.cleanupCreatedChunks(ids);
+            }
+            masterDao.cleanupTransactions(ids);
+        }
+
+        // abort failed transactions
+        masterDao.abortTransactions(transactionManager.activeTransactions(), clock.millis());
+
+        // cleanup failed transactions
+        now = clock.millis();
+        long maxEndTime = now - FAILED_TRANSACTION_AGE.toMillis();
+        long purgeTime = now + FAILED_TRANSACTION_AGE.toMillis();
```

Comment:
So we delay 1 hour to let ChunkCleaner physically delete. But no delay is still safe, right? 
And our production server encounters the "out of disk space", due to a lot of failed inserts. I am wondering whether we should delete in time, given its safe.

Reply:
This is cleaning up data in the chunk store for failed transactions. It is always safe to delete immediately. However, the write might still be happening on a worker, so we could perform the cleanup too early and miss it, which would leak the file forever.

Note that this is the chunk store, not local storage on workers. We expect that the chunk store has a much larger buffer of free space compared to workers, so we can easily afford to wait an hour.

Reply:
I see. solid